### ISSUE TITLE
Hooking a timer causes invalid memory access error

### ISSUE THREAD

**Issue by Misiur** [pawn-lang/YSI-Includes#27]
Found accidentally by @Larceny-

``` pawn
#include <a_samp>
#define _inc_a_samp
#define _DEBUG 7
#include <YSI\y_timers>
#include <YSI\y_hooks>

main () {
}

task WTF[200]()
{
    print("Hello");
    return 1;
}

hook WTF()
{
    print("Hello hooked!");
    return 1;
}
```

Log: https://gist.github.com/Misiur/d3bea962d4230db96812

Preprocessed:

``` pawn
forward Malloc_main();public Malloc_main() {
}

@yT_WTF(g,p);@yT_WTF(g,p){static s=-1;return _Timer_I(# WTF,200,g,s);} WTF();public WTF()
{
    print("Hello");
    return 1;
}

@yH_WTF@003();@yH_WTF@003()
{
    print("Hello hooked!");
    return 1;
}
```

**Comment by Misiur** [comment]
Only thing I noticed so far, that last `AMX_GetEntryPrefix` result in hook (-201712) 
https://gist.github.com/Misiur/d3bea962d4230db96812#file-server_log-txt-L225-L244
is different than -201700 fetched later
https://gist.github.com/Misiur/d3bea962d4230db96812#file-server_log-txt-L610-L611

**Comment by Y-Less** [comment]
I have never figured out why hooking timers doesn't work! At one point I even involved kalcor in the process and got a copy of the SA:MP timers source code just to check if that was doing something odd (it wasn't). That change in AMX_GetEntryPrefix result may or may not mean something - YSI moves publics about in many ways.

When functions are hooked, they are all combined to be called from a single function. That function becomes the main public entry point and is what the AMX header now points to. I would actually thus say that it might be a good thing that that result changed as it shows that the function pointer moved from the real implementation to the y_hooks generated call stub.

Plus, the AMX header itself is manipulated, once hook functions have been generated, all the header references to the originals are unneeded and removed, and the resulting header is compressed to save time when the VM calls a public. A side-effect of this is that you can't do CallLocalFunction("@yH_ScriptInit@015", ""); to call your hook directly, but since hook names are a non-constant implementation detail that's not a huge problem.

Anyway, long story short, I've never got public timer hooks working (even though there should be no reason why not given that they are just normal public functions as far as the VM is concerned), and as a result never even TRIED hooking timer functions.

**Comment by Y-Less** [comment]
I've had an idea, one I've never had before:

``` pawn
@emit HALT 1
```

Instead of generating a real hook stub for a timer (don't know how you will determine it is a timer - I suggest just looking for one specific name), just generate that one instruction.  IF the timer stub gets hit (regular timer, not y_timers) the VM will crash, but at least you will absolutely see it get hit with a single instruction.

**Comment by Misiur** [comment]
I did

``` pawn
#include <a_samp>
#define _inc_a_samp
#define _DEBUG 7
#include <YSI\y_hooks>

main () {
    SetTimer("WTF", 100, 0);
}

forward WTF();
public WTF()
{
    print("Hello");
    return 1;
}

hook WTF()
{
    print("Hello hooked!");
    return 1;
}
```

``` pawn
new size;
if (!strcmp(name, "WTF")) {
    size = Hooks_WriteFunctionTest(hooks, count, Hooks_GetDefaultReturn(name));
} else {
    size = Hooks_WriteFunction(hooks, count, Hooks_GetDefaultReturn(name));
}
```

``` pawn
_Y_HOOKS_STATIC Hooks_WriteFunctionTest(const pointers[], const size, const ret = 1, const skipable = true)
{
    new
        bool:multiple = size != 1,
        base = (AMX_HEADER_COD - AMX_BASE_ADDRESS) + AMX_REAL_ADDRESS,
        ctx[AsmContext];
    // Make sure the underlying system doesn't change without us.  Now supported
    // natively.
    CGen_UseCodeSpace(ctx);
    P:3("Test called!");

    // Start of the function.
    @emit HALT 1

    // Return the number of bytes written.
    return ctx[AsmContext_buffer_offset];
}
```

(I'm not certain what I'm doing, but I hope there isn't a difference if whole reserved space isn't written to).

Log: https://gist.github.com/Misiur/181ca64b10d5792a97a2
It seems it got hit.

**Comment by Misiur** [comment]
I found the most peculiar thing. I just skipped first item of `pointers` in `Hooks_WriteFunction`, and it works. Tho it skips the last hook. If you have 2 hooks, 1 (first one declared) will be called.

**Comment by Misiur** [comment]
Even not exactly like that. It only drops the first hook (the original function is called after all hooks).

Setting `skippable` argument to false also works.

**Comment by Misiur** [comment]
Ok, it seems I fooled myself. It seems it works without touching anything, but you have to use `SetTimerEx` with at least 1 parameter (even if the function doesn't accept any parameters!)

**Comment by Misiur** [comment]
Ok, last observation for today using (normal SetTimer call)

``` pawn
_Y_HOOKS_STATIC Hooks_WriteFunctionTest(const pointers[], const size, const ret = 1, const skipable = false)
{
    new
        bool:multiple = size != 1,
        base = (AMX_HEADER_COD - AMX_BASE_ADDRESS) + AMX_REAL_ADDRESS,
        ctx[AsmContext];
    // Make sure the underlying system doesn't change without us.  Now supported
    // natively.
    CGen_UseCodeSpace(ctx);

    // Start of the function.
    @emit PROC              // 1
    @emit RETN

    // Return the number of bytes written.
    return ctx[AsmContext_buffer_offset];
}
```

```
[21:16:50] Hooks_GenerateCode WTF                              called:      0      0 00000118 -156120
...
[21:16:51] Hello hooked!
```

If this isn't some corrupted stack quirk, then I'm completely clueless.

**Comment by Misiur** [comment]
``` pawn
#include <a_samp>
#define _inc_a_samp
// #define _DEBUG 7
#include <YSI\y_hooks>

main () {
    // CallLocalFunction("WTF", "");
    SetTimerEx("WTF", 200, 0, "s", "Invalid");
}

forward WTF();
public WTF()
{
    print("Hello");
    return 1;
}

hook WTF()
{
    print("Hello hooked!");
    return 1;
}
```

```
[14:50:02] Number of vehicle models: 0
[14:50:02] Hello hooked!
[14:50:02] Hello
```

Pasting that here to keep track

**Comment by Y-Less** [comment]
I don't know why this happens, but purely by fluke I discovered that the code I wrote to support y_hooks in the JIT also works for timers.  So as long as I only use that method and not the other one, this is solved I guess!

**Comment by Y-Less** [comment]
That then helped me narrow down the real issue.  Still not totally sure why timers seemed to trigger it - specifically why timers with NO PARAMETERS did, while other callbacks with no parameters didn't, but regardless I've added a jump over the parameter copying code for the 0 parameter case, so no allocations or copies are (an optimisation for that case as well as a fix, so woo).
